sklearn/naive_bayes.py

```diff
@@ -1425,3 +1437,591 @@ def _joint_log_likelihood(self, X):
             jll += self.feature_log_prob_[i][:, indices].T
         total_ll = jll + self.class_log_prior_
         return total_ll
+
+
+def _fit_one(estimator, X, y, message_clsname="", message=None, **fit_params):
+    """Call ``estimator.fit`` and print elapsed time message.
+
+    See :func:`sklearn.pipeline._fit_one`.
+    """
+    with _print_elapsed_time(message_clsname, message):
+        return estimator.fit(X, y, **fit_params)
+
+
+def _partial_fit_one(estimator, X, y, message_clsname="", message=None, **fit_params):
+    """Call ``estimator.partial_fit`` and print elapsed time message.
+
+    See :func:`sklearn.pipeline._fit_one`.
+    """
+    with _print_elapsed_time(message_clsname, message):
+        return estimator.partial_fit(X, y, **fit_params)
+
+
+def _jll_one(estimator, X):
+    """Call ``estimator._joint_log_likelihood``.
+
+    See :func:`sklearn.pipeline._transform_one`.
+    """
+    return estimator._joint_log_likelihood(estimator._check_X(X))
+
+
+class ColumnwiseNB(_BaseNB, _BaseComposition):
+    """
+    Column-wise Naive Bayes meta-estimator.
+
+    This estimator combines various naive Bayes estimators by applying them
+    to different column subsets of the input and joining their predictions
+    according to the naive Bayes assumption. This is useful when features are
+    heterogeneous and follow different kinds of distributions.
+
+    Parameters
+    ----------
+    nb_estimators : list of tuples
+        List of (name, nb_estimator, columns) tuples specifying the naive Bayes
+        estimators to be combined into a single naive Bayes meta-estimator.
+
+        name : str
+            Name of the naive Bayes estimator. Like in
+            :class:`~sklearn.pipeline.Pipeline`,
+            :class:`~sklearn.pipeline.FeatureUnion`,
+            and :class:`~sklearn.compose.ColumnTransformer`, this allows the
+            subestimator and its parameters to be set using :term:`set_params`
+            and searched in grid search.
+        nb_estimator : estimator
+            The estimator must support :term:`fit` or :term:`partial_fit`,
+            depending on how the meta-estimator is fitted. In addition, the
+            estimator must support ``_joint_log_likelihood`` method, which
```

Comment:
You should not expose a private API. To me you should only request that the NB estimators have a predict_proba ($P(x|y)$) and a `.priors_` or `class_prior_ ` $P(y)$ attribute unless I miss something.

Reply:
I am afraid I have no choice other than to use ``_joint_log_likelihood``, because `predict_proba` returns $P(y|x)$, and not $P(x|y)$. I have double checked how `predic_proba` is computed in `_BaseNB` and also that every row of `predic_proba[i,j]` sums to 1 (in general, we expect $\sum_j P(y=j|\mathbf{x}^i)=1$ but $\sum_j P(\mathbf{x}^i|y=j)\neq1$). As far as I understand, the only place naive Bayes classifiers expose their generative part is `_joint_log_likelihood`. It is somewhat logical that $P(x,y)$ or $P(x|y)$ (which we want) are not in the public API, because the user and "outside" sklearn are typically concerned only about the discriminative function of classifiers.

I don't feel good about referring to `_joint_log_likelihood` in the documentation either, but I had to explain what is acceptable as `nb_estimator`. The only alternative I saw was to mention the private class `_BaseNB`, which seemed less pythonic to me.

Reply:
Yes, I think the [normalisation in `predict_log_proba`](https://github.com/scikit-learn/scikit-learn/blob/fc9be34b2011bc48111b7e750c18ae364d8cc5d0/sklearn/naive_bayes.py#L103-L105) makes this tricky. I had imagined keeping `_joint_log_likelihood` as a "protected" API... but we could expose it as public? Or we could rename it to `decision_function` since it is functionally very similar to that method in LogisticRegression etc. Wdyt, @agramfort ?

Reply:
Should you decide to make it public, I can create a PR. I think it would be best to add a public twin `joint_log_likelihood` to `_BaseNB`, which would do checks before `return self._joint_log_likelihood(X)`. This way, we don't have to modify existing naive Bayes estimators. The meta-estimator will call the public `joint_log_likelihood`. The user may provide existing or custom NBs as sub-estimators. Whether a custom sub-estimator implements the checks in `joint_log_likelihood`, does not seem to matter. 

By the way, I had an issue with `BernoulliNB`, which binarises `X` in `_check_X(X)` at predict. So the meta-estimator had to pull `X` through [it here](https://github.com/scikit-learn/scikit-learn/blob/e1449a0d3660bd33356da03c0ddf010ba740cfa8/sklearn/naive_bayes.py#L1632). This would now become the responsibility of the public `joint_log_likelihood`, not of the meta-estimator.

Reply:
PR created: #23683

Reply:
> PR created: #23683

It is has been merged, and all changes are reflected here. With the new `predict_joint_log_proba`, no more references to private API. The docstring now reads:

```
        nb_estimator : estimator
            The estimator must support :term:`fit` or :term:`partial_fit`,
            depending on how the meta-estimator is fitted. In addition, the
            estimator must support ``predict_joint_log_proba`` method, which
            takes :term:`X` of shape (n_samples, n_features) and returns a
            numpy array of shape (n_samples, n_classes) containing joint
            log-probabilities, ``log P(x,y)`` for each sample point and class.
```